pkg/actuators/machineset/controller_test.go

```diff
@@ -0,0 +1,359 @@
+/*
+Copyright 2021.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package machineset
+
+import (
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	gtypes "github.com/onsi/gomega/types"
+	ibmclient "github.com/openshift/cluster-api-provider-ibmcloud/pkg/actuators/client"
+	mockibm "github.com/openshift/cluster-api-provider-ibmcloud/pkg/actuators/client/mock"
+	ibmcloudproviderv1 "github.com/openshift/cluster-api-provider-ibmcloud/pkg/apis/ibmcloudprovider/v1beta1"
+	machinev1 "github.com/openshift/machine-api-operator/pkg/apis/machine/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+)
+
+var _ = Describe("Reconciler", func() {
+	var c client.Client
+	var stopMgr context.CancelFunc
+	var fakeRecorder *record.FakeRecorder
+	var namespace *corev1.Namespace
+	var mockCtrl *gomock.Controller
+
+	// Mock
+	mockCtrl = gomock.NewController(GinkgoT())
+	// defer mockCtrl.Finish()
+	mockIBMClient := mockibm.NewMockClient(mockCtrl)
+
+	gomock.InOrder(
+		mockIBMClient.EXPECT().InstanceGetProfile(gomock.Any()).Return(false, errors.New("the provided instance profile ID does not exist")),
+		mockIBMClient.EXPECT().InstanceGetProfile(gomock.Any()).Return(true, nil),
+		mockIBMClient.EXPECT().InstanceGetProfile(gomock.Any()).Return(true, nil),
+		mockIBMClient.EXPECT().InstanceGetProfile(gomock.Any()).Return(true, nil),
+		mockIBMClient.EXPECT().InstanceGetProfile(gomock.Any()).Return(true, nil),
+		mockIBMClient.EXPECT().InstanceGetProfile(gomock.Any()).Return(true, nil),
+		mockIBMClient.EXPECT().InstanceGetProfile(gomock.Any()).Return(true, nil),
+		mockIBMClient.EXPECT().InstanceGetProfile(gomock.Any()).Return(false, errors.New("the provided instance profile ID does not exist")),
```

Comment:
There's no guarantee in a test suite like this that each test only calls the endpoint once, this will be fragile if the controller requeues or calls multiple times.

Would be better to set the inputs and outputs to the mock rather than rely on ordering IMO

Reply:
Agree, made the changes.